ccl/changefeedccl: TestSchemachangeDoesNotBreakSinklessFeed failed

ccl/changefeedccl.TestSchemachangeDoesNotBreakSinklessFeed [failed](https://tanzanite.cluster.engflow.com/invocations/default/bc3a5ff8-2bb0-4df7-8040-56a3ef576797?testReportRun=3&testReportShard=16&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [49e62ff4c0215218fbfc12c2ab3cf3429f0315a3](https://github.com/cockroachdb/cockroach/commits/49e62ff4c0215218fbfc12c2ab3cf3429f0315a3):


```
test logs left over in: outputs.zip/logTestSchemachangeDoesNotBreakSinklessFeed2617045424
--- FAIL: TestSchemachangeDoesNotBreakSinklessFeed (19.44s)
=== RUN   TestSchemachangeDoesNotBreakSinklessFeed/sinkless
    helpers_test.go:883: making server as secondary tenant
    helpers_test.go:962: making sinkless feed factory
    helpers_test.go:1027: pgURL sinkless SinklessFeedUser
    helpers_test.go:1027: pgURL sinkless root
    testfeed_test.go:271: sinkless feed creating changefeed: CREATE CHANGEFEED FOR TABLE mytable
    changefeed_test.go:9028: 
        	Error Trace:	pkg/ccl/changefeedccl/helpers_test.go:198
        	            				pkg/ccl/changefeedccl/helpers_test.go:273
        	            				pkg/ccl/changefeedccl/changefeed_test.go:9028
        	            				pkg/ccl/changefeedccl/helpers_test.go:1134
        	            				pkg/ccl/changefeedccl/helpers_test.go:1169
        	Error:      	Received unexpected error:
        	            	expected message
        	            	(1) assertion failure
        	            	Wraps: (2) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.readNextMessages
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:118
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBaseErr
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:209
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBase.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:201
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.withTimeout.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:266
        	            	  | github.com/cockroachdb/cockroach/pkg/util/timeutil.RunWithTimeout
        	            	  | 	pkg/util/timeutil/timeout.go:28
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.withTimeout
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:262
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBase
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:199
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloads
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:273
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestSchemachangeDoesNotBreakSinklessFeed.func1
        	            	  | 	pkg/ccl/changefeedccl/changefeed_test.go:9028
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamed.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:1134
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamedWithSystem.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:1169
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1689
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1695
        	            	Wraps: (3) expected message
        	            	Error types: (1) *assert.withAssertionFailure (2) *withstack.withStack (3) *errutil.leafError
        	Test:       	TestSchemachangeDoesNotBreakSinklessFeed/sinkless
    testfeed_test.go:278: closing sinkless feed
    --- FAIL: TestSchemachangeDoesNotBreakSinklessFeed/sinkless (19.43s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=3</code>
 - <code>shard=16</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSchemachangeDoesNotBreakSinklessFeed.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-46823

ccl/changefeedccl.TestSchemachangeDoesNotBreakSinklessFeed [failed](https://tanzanite.cluster.engflow.com/invocations/default/dec99969-f725-4710-9ff3-9197dfd77f61?testReportRun=3&testReportShard=16&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [529a310186182ad0ebce700708f05193f629217b](https://github.com/cockroachdb/cockroach/commits/529a310186182ad0ebce700708f05193f629217b):


```
test logs left over in: outputs.zip/logTestSchemachangeDoesNotBreakSinklessFeed3609479831
--- FAIL: TestSchemachangeDoesNotBreakSinklessFeed (17.27s)
=== RUN   TestSchemachangeDoesNotBreakSinklessFeed/sinkless
    helpers_test.go:916: making server as secondary tenant
    helpers_test.go:995: making sinkless feed factory
    helpers_test.go:1060: pgURL sinkless SinklessFeedUser
    helpers_test.go:1060: pgURL sinkless root
    testfeed_test.go:271: sinkless feed creating changefeed: CREATE CHANGEFEED FOR TABLE mytable
    changefeed_test.go:9184: 
        	Error Trace:	pkg/ccl/changefeedccl/helpers_test.go:219
        	            				pkg/ccl/changefeedccl/helpers_test.go:299
        	            				pkg/ccl/changefeedccl/changefeed_test.go:9184
        	            				pkg/ccl/changefeedccl/helpers_test.go:1167
        	            				pkg/ccl/changefeedccl/helpers_test.go:1202
        	Error:      	Received unexpected error:
        	            	expected message
        	            	(1) assertion failure
        	            	Wraps: (2) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.readNextMessages
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:118
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBaseErr
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:235
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBase.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:222
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.withTimeout.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:292
        	            	  | github.com/cockroachdb/cockroach/pkg/util/timeutil.RunWithTimeout
        	            	  | 	pkg/util/timeutil/timeout.go:28
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.withTimeout
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:288
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBase
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:220
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloads
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:299
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestSchemachangeDoesNotBreakSinklessFeed.func1
        	            	  | 	pkg/ccl/changefeedccl/changefeed_test.go:9184
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamed.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:1167
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamedWithSystem.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:1202
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1689
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1695
        	            	Wraps: (3) expected message
        	            	Error types: (1) *assert.withAssertionFailure (2) *withstack.withStack (3) *errutil.leafError
        	Test:       	TestSchemachangeDoesNotBreakSinklessFeed/sinkless
    testfeed_test.go:278: closing sinkless feed
    --- FAIL: TestSchemachangeDoesNotBreakSinklessFeed/sinkless (17.26s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=3</code>
 - <code>shard=16</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #141355 ccl/changefeedccl: TestSchemachangeDoesNotBreakSinklessFeed failed [A-cdc C-test-failure O-robot P-2 T-cdc X-duplicate branch-release-25.1]
- #141128 ccl/changefeedccl: TestSchemachangeDoesNotBreakSinklessFeed failed [A-cdc C-test-failure O-robot T-cdc X-duplicate branch-release-24.2]
- #140796 ccl/changefeedccl: TestSchemachangeDoesNotBreakSinklessFeed failed [A-cdc C-test-failure O-robot P-2 T-cdc X-duplicate branch-release-25.1.0-rc]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSchemachangeDoesNotBreakSinklessFeed.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

ccl/changefeedccl.TestSchemachangeDoesNotBreakSinklessFeed [failed](https://tanzanite.cluster.engflow.com/invocations/default/bff06dca-8911-4aca-a64b-3dc43409a2cc?testReportRun=3&testReportShard=16&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [cbb292c487ea7413772a5c3c349f93467264c190](https://github.com/cockroachdb/cockroach/commits/cbb292c487ea7413772a5c3c349f93467264c190):


```
test logs left over in: outputs.zip/logTestSchemachangeDoesNotBreakSinklessFeed178790422
--- FAIL: TestSchemachangeDoesNotBreakSinklessFeed (18.89s)
=== RUN   TestSchemachangeDoesNotBreakSinklessFeed/sinkless
    helpers_test.go:917: making server as secondary tenant
    helpers_test.go:996: making sinkless feed factory
    helpers_test.go:1061: pgURL sinkless SinklessFeedUser
    helpers_test.go:1061: pgURL sinkless root
    testfeed_test.go:271: sinkless feed creating changefeed: CREATE CHANGEFEED FOR TABLE mytable
    changefeed_test.go:9209: 
        	Error Trace:	pkg/ccl/changefeedccl/helpers_test.go:220
        	            				pkg/ccl/changefeedccl/helpers_test.go:300
        	            				pkg/ccl/changefeedccl/changefeed_test.go:9209
        	            				pkg/ccl/changefeedccl/helpers_test.go:1168
        	            				pkg/ccl/changefeedccl/helpers_test.go:1203
        	Error:      	Received unexpected error:
        	            	expected message
        	            	(1) assertion failure
        	            	Wraps: (2) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.readNextMessages
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:119
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBaseErr
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:236
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBase.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:223
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.withTimeout.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:293
        	            	  | github.com/cockroachdb/cockroach/pkg/util/timeutil.RunWithTimeout
        	            	  | 	pkg/util/timeutil/timeout.go:28
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.withTimeout
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:289
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBase
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:221
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloads
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:300
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestSchemachangeDoesNotBreakSinklessFeed.func1
        	            	  | 	pkg/ccl/changefeedccl/changefeed_test.go:9209
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamed.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:1168
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamedWithSystem.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:1203
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1689
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1695
        	            	Wraps: (3) expected message
        	            	Error types: (1) *assert.withAssertionFailure (2) *withstack.withStack (3) *errutil.leafError
        	Test:       	TestSchemachangeDoesNotBreakSinklessFeed/sinkless
    testfeed_test.go:278: closing sinkless feed
    --- FAIL: TestSchemachangeDoesNotBreakSinklessFeed/sinkless (18.89s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=3</code>
 - <code>shard=16</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #141355 ccl/changefeedccl: TestSchemachangeDoesNotBreakSinklessFeed failed [A-cdc C-test-failure O-robot P-2 T-cdc X-duplicate branch-release-25.1]
- #141128 ccl/changefeedccl: TestSchemachangeDoesNotBreakSinklessFeed failed [A-cdc C-test-failure O-robot P-2 T-cdc X-duplicate branch-release-24.2]
- #140796 ccl/changefeedccl: TestSchemachangeDoesNotBreakSinklessFeed failed [A-cdc C-test-failure O-robot P-2 T-cdc X-duplicate branch-release-25.1.0-rc]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSchemachangeDoesNotBreakSinklessFeed.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

ccl/changefeedccl.TestSchemachangeDoesNotBreakSinklessFeed [failed](https://tanzanite.cluster.engflow.com/invocations/default/d5542db9-894f-48f5-8954-6bb4f7e4bed8?testReportRun=1&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [478d65a28755b45c2952cc755a0cc8ae466514a7](https://github.com/cockroachdb/cockroach/commits/478d65a28755b45c2952cc755a0cc8ae466514a7):


```
test logs left over in: outputs.zip/logTestSchemachangeDoesNotBreakSinklessFeed1015437438
--- FAIL: TestSchemachangeDoesNotBreakSinklessFeed (16.57s)
=== RUN   TestSchemachangeDoesNotBreakSinklessFeed/sinkless
    helpers_test.go:945: making server as secondary tenant
    helpers_test.go:1024: making sinkless feed factory
    helpers_test.go:1089: pgURL sinkless SinklessFeedUser
    helpers_test.go:1089: pgURL sinkless root
    testfeed_test.go:271: sinkless feed creating changefeed: CREATE CHANGEFEED FOR TABLE mytable
    changefeed_test.go:9267: 
        	Error Trace:	pkg/ccl/changefeedccl/helpers_test.go:220
        	            				pkg/ccl/changefeedccl/helpers_test.go:300
        	            				pkg/ccl/changefeedccl/changefeed_test.go:9267
        	            				pkg/ccl/changefeedccl/helpers_test.go:1196
        	            				pkg/ccl/changefeedccl/helpers_test.go:1231
        	Error:      	Received unexpected error:
        	            	expected message
        	            	(1) assertion failure
        	            	Wraps: (2) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.readNextMessages
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:119
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBaseErr
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:236
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBase.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:223
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.withTimeout.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:293
        	            	  | github.com/cockroachdb/cockroach/pkg/util/timeutil.RunWithTimeout
        	            	  | 	pkg/util/timeutil/timeout.go:28
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.withTimeout
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:289
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBase
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:221
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloads
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:300
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestSchemachangeDoesNotBreakSinklessFeed.func1
        	            	  | 	pkg/ccl/changefeedccl/changefeed_test.go:9267
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamed.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:1196
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamedWithSystem.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:1231
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1690
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1700
        	            	Wraps: (3) expected message
        	            	Error types: (1) *assert.withAssertionFailure (2) *withstack.withStack (3) *errutil.leafError
        	Test:       	TestSchemachangeDoesNotBreakSinklessFeed/sinkless
    testfeed_test.go:278: closing sinkless feed
    --- FAIL: TestSchemachangeDoesNotBreakSinklessFeed/sinkless (16.57s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #141355 ccl/changefeedccl: TestSchemachangeDoesNotBreakSinklessFeed failed [A-cdc C-test-failure O-robot P-2 T-cdc X-duplicate branch-release-25.1]
- #141128 ccl/changefeedccl: TestSchemachangeDoesNotBreakSinklessFeed failed [A-cdc C-test-failure O-robot P-2 T-cdc X-duplicate branch-release-24.2]
- #140796 ccl/changefeedccl: TestSchemachangeDoesNotBreakSinklessFeed failed [A-cdc C-test-failure O-robot P-2 T-cdc X-duplicate branch-release-25.1.0-rc]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSchemachangeDoesNotBreakSinklessFeed.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

ccl/changefeedccl.TestSchemachangeDoesNotBreakSinklessFeed [failed](https://tanzanite.cluster.engflow.com/invocations/default/e467f831-1a48-41f1-aad3-7d1b861f0b0e?testReportRun=2&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [59dedfedb416900dfbdbbcc6ef5c8dfc087b7bcf](https://github.com/cockroachdb/cockroach/commits/59dedfedb416900dfbdbbcc6ef5c8dfc087b7bcf):


```
test logs left over in: outputs.zip/logTestSchemachangeDoesNotBreakSinklessFeed2245276862
--- FAIL: TestSchemachangeDoesNotBreakSinklessFeed (15.68s)
=== RUN   TestSchemachangeDoesNotBreakSinklessFeed/sinkless
    helpers_test.go:1031: making server as secondary tenant
    helpers_test.go:1110: making sinkless feed factory
    helpers_test.go:1175: pgURL sinkless SinklessFeedUser
    helpers_test.go:1175: pgURL sinkless root
    testfeed_test.go:291: sinkless feed creating changefeed: CREATE CHANGEFEED FOR TABLE mytable
    changefeed_test.go:9859: 
        	Error Trace:	pkg/ccl/changefeedccl/helpers_test.go:259
        	            				pkg/ccl/changefeedccl/helpers_test.go:354
        	            				pkg/ccl/changefeedccl/changefeed_test.go:9859
        	            				pkg/ccl/changefeedccl/helpers_test.go:1282
        	            				pkg/ccl/changefeedccl/helpers_test.go:1317
        	Error:      	Received unexpected error:
        	            	expected message
        	            	(1) assertion failure
        	            	Wraps: (2) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.readNextMessages
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:122
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBaseErr
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:276
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBase.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:262
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.withTimeout.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:347
        	            	  | github.com/cockroachdb/cockroach/pkg/util/timeutil.RunWithTimeout
        	            	  | 	pkg/util/timeutil/timeout.go:28
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.withTimeout
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:343
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBase
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:260
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloads
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:354
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestSchemachangeDoesNotBreakSinklessFeed.func1
        	            	  | 	pkg/ccl/changefeedccl/changefeed_test.go:9859
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamed.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:1282
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamedWithSystem.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:1317
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1690
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1700
        	            	Wraps: (3) expected message
        	            	Error types: (1) *assert.withAssertionFailure (2) *withstack.withStack (3) *errutil.leafError
        	Test:       	TestSchemachangeDoesNotBreakSinklessFeed/sinkless
    testfeed_test.go:298: closing sinkless feed
--- FAIL: TestSchemachangeDoesNotBreakSinklessFeed/sinkless (15.67s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=2</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #141355 ccl/changefeedccl: TestSchemachangeDoesNotBreakSinklessFeed failed [A-cdc C-test-failure O-robot P-2 T-cdc X-duplicate branch-release-25.1]
- #141128 ccl/changefeedccl: TestSchemachangeDoesNotBreakSinklessFeed failed [A-cdc C-test-failure O-robot P-2 T-cdc X-duplicate branch-release-24.2]
- #140796 ccl/changefeedccl: TestSchemachangeDoesNotBreakSinklessFeed failed [A-cdc C-test-failure O-robot P-2 T-cdc X-duplicate branch-release-25.1.0-rc]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSchemachangeDoesNotBreakSinklessFeed.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

ccl/changefeedccl.TestSchemachangeDoesNotBreakSinklessFeed [failed](https://tanzanite.cluster.engflow.com/invocations/default/45b66acf-17c2-437f-a437-6eb6efb34726?testReportRun=1&testReportShard=7&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on master @ [7d913b352d0cd31d267cc7da455446d0cb801b59](https://github.com/cockroachdb/cockroach/commits/7d913b352d0cd31d267cc7da455446d0cb801b59):


```
test logs left over in: outputs.zip/logTestSchemachangeDoesNotBreakSinklessFeed1469960438
--- FAIL: TestSchemachangeDoesNotBreakSinklessFeed (15.85s)
=== RUN   TestSchemachangeDoesNotBreakSinklessFeed/sinkless
    helpers_test.go:1031: making server as secondary tenant
    helpers_test.go:1110: making sinkless feed factory
    helpers_test.go:1175: pgURL sinkless SinklessFeedUser
    helpers_test.go:1175: pgURL sinkless root
    testfeed_test.go:291: sinkless feed creating changefeed: CREATE CHANGEFEED FOR TABLE mytable
    changefeed_test.go:9859: 
        	Error Trace:	pkg/ccl/changefeedccl/helpers_test.go:259
        	            				pkg/ccl/changefeedccl/helpers_test.go:354
        	            				pkg/ccl/changefeedccl/changefeed_test.go:9859
        	            				pkg/ccl/changefeedccl/helpers_test.go:1282
        	            				pkg/ccl/changefeedccl/helpers_test.go:1317
        	Error:      	Received unexpected error:
        	            	expected message
        	            	(1) assertion failure
        	            	Wraps: (2) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.readNextMessages
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:122
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBaseErr
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:276
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBase.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:262
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.withTimeout.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:347
        	            	  | github.com/cockroachdb/cockroach/pkg/util/timeutil.RunWithTimeout
        	            	  | 	pkg/util/timeutil/timeout.go:28
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.withTimeout
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:343
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloadsBase
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:260
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.assertPayloads
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:354
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.TestSchemachangeDoesNotBreakSinklessFeed.func1
        	            	  | 	pkg/ccl/changefeedccl/changefeed_test.go:9859
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamed.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:1282
        	            	  | github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl.cdcTestNamedWithSystem.func1
        	            	  | 	pkg/ccl/changefeedccl/helpers_test.go:1317
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1690
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1700
        	            	Wraps: (3) expected message
        	            	Error types: (1) *assert.withAssertionFailure (2) *withstack.withStack (3) *errutil.leafError
        	Test:       	TestSchemachangeDoesNotBreakSinklessFeed/sinkless
    testfeed_test.go:298: closing sinkless feed
--- FAIL: TestSchemachangeDoesNotBreakSinklessFeed/sinkless (15.85s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=7</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #141355 ccl/changefeedccl: TestSchemachangeDoesNotBreakSinklessFeed failed [A-cdc C-test-failure O-robot P-2 T-cdc X-duplicate branch-release-25.1]
- #141128 ccl/changefeedccl: TestSchemachangeDoesNotBreakSinklessFeed failed [A-cdc C-test-failure O-robot P-2 T-cdc X-duplicate branch-release-24.2]
- #140796 ccl/changefeedccl: TestSchemachangeDoesNotBreakSinklessFeed failed [A-cdc C-test-failure O-robot P-2 T-cdc X-duplicate branch-release-25.1.0-rc]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSchemachangeDoesNotBreakSinklessFeed.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/ccl/changefeedccl/changefeed_processors.go
+++ b/pkg/ccl/changefeedccl/changefeed_processors.go
@@ -446,8 +446,15 @@ func (ca *changeAggregator) startKVFeed(
 	cfg := ca.FlowCtx.Cfg
 	kvFeedMemMon := mon.NewMonitorInheritWithLimit(mon.MakeName("kvFeed"), memLimit, parentMemMon, false /* longLiving */)
 	kvFeedMemMon.StartNoReserved(ctx, parentMemMon)
+
+	var options []kvevent.BlockingBufferOption
+	if ca.knobs.MakeKVFeedToAggregatorBufferKnobs != nil {
+		options = append(options,
+			kvevent.WithBlockingBufferTestingKnobs(ca.knobs.MakeKVFeedToAggregatorBufferKnobs()))
+	}
 	buf := kvevent.NewThrottlingBuffer(
-		kvevent.NewMemBuffer(kvFeedMemMon.MakeBoundAccount(), &cfg.Settings.SV, &ca.metrics.KVFeedMetrics.AggregatorBufferMetricsWithCompat),
+		kvevent.NewMemBuffer(kvFeedMemMon.MakeBoundAccount(), &cfg.Settings.SV,
+			&ca.metrics.KVFeedMetrics.AggregatorBufferMetricsWithCompat, options...),
 		cdcutils.NodeLevelThrottler(&cfg.Settings.SV, &ca.metrics.ThrottleMetrics))
 
 	// KVFeed takes ownership of the kvevent.Writer portion of the buffer, while

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -1822,7 +1822,9 @@ func TestNoStopAfterNonTargetColumnDrop(t *testing.T) {
 		}
 	}
 
-	cdcTest(t, testFn)
+	runWithAndWithoutRegression141453(t, testFn, func(t *testing.T, testFn cdcTestFn) {
+		cdcTest(t, testFn)
+	})
 }
 
 func TestChangefeedProjectionDelete(t *testing.T) {
@@ -2062,7 +2064,9 @@ func TestChangefeedColumnDropsOnTheSameTableWithMultipleFamilies(t *testing.T) {
 		})
 	}
 
-	cdcTest(t, testFn)
+	runWithAndWithoutRegression141453(t, testFn, func(t *testing.T, testFn cdcTestFn) {
+		cdcTest(t, testFn)
+	})
 }
 
 func TestNoStopAfterNonTargetAddColumnWithBackfill(t *testing.T) {
@@ -2106,7 +2110,9 @@ func TestNoStopAfterNonTargetAddColumnWithBackfill(t *testing.T) {
 		}
 	}
 
-	cdcTest(t, testFn)
+	runWithAndWithoutRegression141453(t, testFn, func(t *testing.T, testFn cdcTestFn) {
+		cdcTest(t, testFn)
+	})
 }
 
 func TestChangefeedExternalIODisabled(t *testing.T) {
@@ -2994,7 +3000,7 @@ func TestChangefeedSchemaChangeAllowBackfill(t *testing.T) {
 
 	require.NoError(t, log.SetVModule("kv_feed=2,changefeed_processors=2"))
 
-	testFn := func(t *testing.T, s TestServerWithSystem, f cdctest.TestFeedFactory) {
+	testFn := func(t *testing.T, s TestServer, f cdctest.TestFeedFactory) {
 		sqlDB := sqlutils.MakeSQLRunner(s.DB)
 
 		// Expected semantics:
@@ -3025,7 +3031,7 @@ func TestChangefeedSchemaChangeAllowBackfill(t *testing.T) {
 				`add_column_def: [2]->{"after": {"a": 2}}`,
 			})
 			sqlDB.Exec(t, `ALTER TABLE add_column_def ADD COLUMN b STRING DEFAULT 'd'`)
-			ts := schematestutils.FetchDescVersionModificationTime(t, s.TestServer.Server, `d`, `public`, `add_column_def`, 7)
+			ts := schematestutils.FetchDescVersionModificationTime(t, s.Server, `d`, `public`, `add_column_def`, 7)
 			assertPayloads(t, addColumnDef, []string{
 				fmt.Sprintf(`add_column_def: [1]->{"after": {"a": 1, "b": "d"}, "updated": "%s"}`,
 					ts.AsOfSystemTime()),
@@ -3045,7 +3051,7 @@ func TestChangefeedSchemaChangeAllowBackfill(t *testing.T) {
 				`add_col_comp: [2]->{"after": {"a": 2, "b": 7}}`,
 			})
 			sqlDB.Exec(t, `ALTER TABLE add_col_comp ADD COLUMN c INT AS (a + 10) STORED`)
-			ts := schematestutils.FetchDescVersionModificationTime(t, s.TestServer.Server, `d`, `public`, `add_col_comp`, 7)
+			ts := schematestutils.FetchDescVersionModificationTime(t, s.Server, `d`, `public`, `add_col_comp`, 7)
 			assertPayloads(t, addColComp, []string{
 				fmt.Sprintf(`add_col_comp: [1]->{"after": {"a": 1, "b": 6, "c": 11}, "updated": "%s"}`,
 					ts.AsOfSystemTime()),
@@ -3065,7 +3071,7 @@ func TestChangefeedSchemaChangeAllowBackfill(t *testing.T) {
 				`drop_column: [2]->{"after": {"a": 2, "b": "2"}}`,
 			})
 			sqlDB.Exec(t, `ALTER TABLE drop_column DROP COLUMN b`)
-			ts := schematestutils.FetchDescVersionModificationTime(t, s.TestServer.Server, `d`, `public`, `drop_column`, 2)
+			ts := schematestutils.FetchDescVersionModificationTime(t, s.Server, `d`, `public`, `drop_column`, 2)
 
 			// Backfill for DROP COLUMN b.
 			assertPayloads(t, dropColumn, []string{
@@ -3118,9 +3124,9 @@ func TestChangefeedSchemaChangeAllowBackfill(t *testing.T) {
 			// the 7th step (version 15). Finally, when adding column d, it goes from 17->25 ith the schema change
 			// being visible at the 7th step (version 23).
 			// TODO(#142936): Investigate if this descriptor version hardcoding is sound.
-			dropTS := schematestutils.FetchDescVersionModificationTime(t, s.TestServer.Server, `d`, `public`, `multiple_alters`, 2)
-			addTS := schematestutils.FetchDescVersionModificationTime(t, s.TestServer.Server, `d`, `public`, `multiple_alters`, 15)
-			addTS2 := schematestutils.FetchDescVersionModificationTime(t, s.TestServer.Server, `d`, `public`, `multiple_alters`, 23)
+			dropTS := schematestutils.FetchDescVersionModificationTime(t, s.Server, `d`, `public`, `multiple_alters`, 2)
+			addTS := schematestutils.FetchDescVersionModificationTime(t, s.Server, `d`, `public`, `multiple_alters`, 15)
+			addTS2 := schematestutils.FetchDescVersionModificationTime(t, s.Server, `d`, `public`, `multiple_alters`, 23)
 
 			assertPayloads(t, multipleAlters, []string{
 				fmt.Sprintf(`multiple_alters: [1]->{"after": {"a": 1}, "updated": "%s"}`, dropTS.AsOfSystemTime()),
@@ -3133,7 +3139,9 @@ func TestChangefeedSchemaChangeAllowBackfill(t *testing.T) {
 		})
 	}
 
-	cdcTestWithSystem(t, testFn)
+	runWithAndWithoutRegression141453(t, testFn, func(t *testing.T, testFn cdcTestFn) {
+		cdcTest(t, testFn)
+	})
 
 	log.FlushFiles()
 	entries, err := log.FetchEntriesFromFiles(0, math.MaxInt64, 1,
@@ -3377,8 +3385,6 @@ func TestChangefeedSingleColumnFamilySchemaChanges(t *testing.T) {
 	testFn := func(t *testing.T, s TestServer, f cdctest.TestFeedFactory) {
 		sqlDB := sqlutils.MakeSQLRunner(s.DB)
 
-		_ = maybeDisableDeclarativeSchemaChangesForTest(t, sqlDB)
-
 		// Table with 2 column families.
 		sqlDB.Exec(t, `CREATE TABLE foo (a INT PRIMARY KEY, b STRING, c STRING, FAMILY most (a,b), FAMILY rest (c))`)
 		sqlDB.Exec(t, `INSERT INTO foo values (0, 'dog', 'cat')`)
@@ -3405,9 +3411,11 @@ func TestChangefeedSingleColumnFamilySchemaChanges(t *testing.T) {
 		waitForSchemaChange(t, sqlDB, `ALTER TABLE foo DROP column c`)
 		requireErrorSoon(context.Background(), t, fooRest,
 			regexp.MustCompile(`CHANGEFEED targeting nonexistent or removed column family rest of table foo`))
-
 	}
-	cdcTest(t, testFn)
+
+	runWithAndWithoutRegression141453(t, testFn, func(t *testing.T, testFn cdcTestFn) {
+		cdcTest(t, testFn)
+	}, withMaybeUseLegacySchemaChanger())
 }
 
 func TestChangefeedEachColumnFamilySchemaChanges(t *testing.T) {
@@ -3418,8 +3426,6 @@ func TestChangefeedEachColumnFamilySchemaChanges(t *testing.T) {
 
 	testFn := func(t *testing.T, s TestServer, f cdctest.TestFeedFactory) {
 		sqlDB := sqlutils.MakeSQLRunner(s.DB)
-
-		_ = maybeDisableDeclarativeSchemaChangesForTest(t, sqlDB)
 
 		// Table with 2 column families.
 		sqlDB.Exec(t, `CREATE TABLE foo (a INT PRIMARY KEY, b STRING, c STRING, FAMILY f1 (a,b), FAMILY f2 (c))`)
@@ -3447,7 +3453,10 @@ func TestChangefeedEachColumnFamilySchemaChanges(t *testing.T) {
 			`foo.f3: [0]->{"after": {"e": "hello"}}`,
 		})
 	}
-	cdcTest(t, testFn)
+
+	runWithAndWithoutRegression141453(t, testFn, func(t *testing.T, testFn cdcTestFn) {
+		cdcTest(t, testFn)
+	}, withMaybeUseLegacySchemaChanger())
 }
 
 func TestCoreChangefeedRequiresSelectPrivilege(t *testing.T) {
@@ -9865,7 +9874,9 @@ func TestSchemachangeDoesNotBreakSinklessFeed(t *testing.T) {
 		})
 	}
 
-	cdcTest(t, testFn, feedTestForceSink("sinkless"))
+	runWithAndWithoutRegression141453(t, testFn, func(t *testing.T, testFn cdcTestFn) {
+		cdcTest(t, testFn, feedTestForceSink("sinkless"))
+	})
 }
 
 func TestChangefeedKafkaMessageTooLarge(t *testing.T) {

--- a/pkg/ccl/changefeedccl/helpers_test.go
+++ b/pkg/ccl/changefeedccl/helpers_test.go
@@ -21,6 +21,7 @@ import (
 	"sort"
 	"strconv"
 	"strings"
+	"sync/atomic"
 	"testing"
 	"time"
 
@@ -29,6 +30,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/cdctest"
 	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/changefeedbase"
 	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kcjsonschema"
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvevent"
 	_ "github.com/cockroachdb/cockroach/pkg/ccl/multiregionccl" // allow locality-related mutations
 	"github.com/cockroachdb/cockroach/pkg/ccl/multiregionccl/multiregionccltestutils"
 	_ "github.com/cockroachdb/cockroach/pkg/ccl/partitionccl"
@@ -69,7 +71,7 @@ func maybeDisableDeclarativeSchemaChangesForTest(t testing.TB, sqlDB *sqlutils.S
 	if disable {
 		t.Log("using legacy schema changer")
 		sqlDB.Exec(t, "SET use_declarative_schema_changer='off'")
-		sqlDB.Exec(t, "SET CLUSTER SETTING  sql.defaults.use_declarative_schema_changer='off'")
+		sqlDB.Exec(t, "SET CLUSTER SETTING sql.defaults.use_declarative_schema_changer='off'")
 	}
 	return disable
 }
@@ -1628,4 +1630,111 @@ func checkSchema(actual []cdctest.TestFeedMessage) error {
 		}
 	}
 	return nil
+}
+
+type regression141453Options struct {
+	maybeUseLegacySchemaChanger bool
+}
+
+type regression141453Option func(*regression141453Options)
+
+func withMaybeUseLegacySchemaChanger() regression141453Option {
+	return func(opts *regression141453Options) {
+		opts.maybeUseLegacySchemaChanger = true
+	}
+}
+
+// runWithAndWithoutRegression141453 runs the test both with and without testing
+// knobs that simulate the scenario where a change aggregator encounters a schema
+// change restart but draining the buffer fails so the resolved spans message
+// signaling the restart doesn't get sent to the change frontier.
+func runWithAndWithoutRegression141453(
+	t *testing.T,
+	testFn cdcTestFn,
+	runTestFn func(t *testing.T, testFn cdcTestFn),
+	opts ...regression141453Option,
+) {
+	testutils.RunTrueAndFalse(t, "regression 141453",
+		func(t *testing.T, regression141453 bool) {
+			testFn := func(t *testing.T, s TestServer, f cdctest.TestFeedFactory) {
+				var options regression141453Options
+				for _, opt := range opts {
+					opt(&options)
+				}
+
+				var useLegacySchemaChanger bool
+				if options.maybeUseLegacySchemaChanger {
+					sqlDB := sqlutils.MakeSQLRunner(s.DB)
+					useLegacySchemaChanger = maybeDisableDeclarativeSchemaChangesForTest(t, sqlDB)
+				}
+
+				// This regression scenario doesn't always happen with the legacy schema changer
+				// because altering the table sometimes results in backfills instead of restarts.
+				if useLegacySchemaChanger || !regression141453 {
+					testFn(t, s, f)
+					return
+				}
+
+				knobs := s.TestingKnobs.
+					DistSQL.(*execinfra.TestingKnobs).
+					Changefeed.(*TestingKnobs)
+
+				// We force the regression scenario to happen by:
+				// 1. Blocking popping from the kv feed to change aggregator buffer
+				//    before we add the restart resolved span boundary message.
+				// 2. Canceling the context that Drain uses so that it fails.
+				// 3. Re-allowing popping after the buffer is closed.
+				//
+				// This will ensure that the change aggregator will not be able
+				// to pop (and send) the restart resolved span boundary message
+				// and thus the changefeed should restart due to transient error
+				// (before the expected restart for the schema change).
+				//
+				// Previously, this scenario would incorrectly cause the changefeed
+				// to shut down as if it had completed successfully.
+				//
+				// Note that we only want to make Drain fail once, otherwise the test
+				// will never be able to proceed.
+				var drainFailedOnce atomic.Bool
+				knobs.MakeKVFeedToAggregatorBufferKnobs = func() kvevent.BlockingBufferTestingKnobs {
+					if drainFailedOnce.Load() {
+						return kvevent.BlockingBufferTestingKnobs{}
+					}
+					var blockPop atomic.Bool
+					popCh := make(chan struct{})
+					return kvevent.BlockingBufferTestingKnobs{
+						BeforeAdd: func(ctx context.Context, e kvevent.Event) (context.Context, kvevent.Event) {
+							if e.Type() == kvevent.TypeResolved &&
+								e.Resolved().BoundaryType == jobspb.ResolvedSpan_RESTART {
+								blockPop.Store(true)
+							}
+							return ctx, e
+						},
+						BeforePop: func() {
+							if blockPop.Load() {
+								<-popCh
+							}
+						},
+						BeforeDrain: func(ctx context.Context) context.Context {
+							ctx, cancel := context.WithCancel(ctx)
+							cancel()
+							return ctx
+						},
+						AfterDrain: func(err error) {
+							require.Error(t, err)
+							drainFailedOnce.Store(true)
+						},
+						AfterCloseWithReason: func(err error) {
+							require.NoError(t, err)
+							close(popCh)
+							blockPop.Store(false)
+						},
+					}
+				}
+				testFn(t, s, f)
+				require.True(t, drainFailedOnce.Load())
+			}
+
+			runTestFn(t, testFn)
+		})
 }

--- a/pkg/ccl/changefeedccl/kvevent/BUILD.bazel
+++ b/pkg/ccl/changefeedccl/kvevent/BUILD.bazel
@@ -8,6 +8,7 @@ go_library(
         "chunked_event_queue.go",
         "event.go",
         "metrics.go",
+        "testing_knobs.go",
         "throttling_buffer.go",
     ],
     importpath = "github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvevent",

--- a/pkg/ccl/changefeedccl/kvevent/blocking_buffer.go
+++ b/pkg/ccl/changefeedccl/kvevent/blocking_buffer.go
@@ -43,6 +43,8 @@ type blockingBuffer struct {
 		canFlush   bool
 		queue      *bufferEventChunkQueue // Queue of added events.
 	}
+
+	knobs BlockingBufferTestingKnobs
 }
 
 // NewMemBuffer returns a new in-memory buffer which will store events.
@@ -50,9 +52,12 @@ type blockingBuffer struct {
 // runs out of space. If ever any entry exceeds the allocatable size of the
 // account, an error will be returned when attempting to buffer it.
 func NewMemBuffer(
-	acc mon.BoundAccount, sv *settings.Values, metrics *PerBufferMetricsWithCompat,
+	acc mon.BoundAccount,
+	sv *settings.Values,
+	metrics *PerBufferMetricsWithCompat,
+	options ...BlockingBufferOption,
 ) Buffer {
-	return newMemBuffer(acc, sv, metrics, nil)
+	return newMemBuffer(acc, sv, metrics, nil, options...)
 }
 
 // TestingNewMemBuffer allows test to construct buffer which will invoked
@@ -71,6 +76,7 @@ func newMemBuffer(
 	sv *settings.Values,
 	metrics *PerBufferMetricsWithCompat,
 	onWaitStart quotapool.OnWaitStartFunc,
+	options ...BlockingBufferOption,
 ) Buffer {
 	const slowAcquisitionThreshold = 5 * time.Second
 
@@ -109,14 +115,31 @@ func newMemBuffer(
 		metrics:      metrics,
 	}
 
+	for _, opt := range options {
+		opt(b)
+	}
+
 	return b
 }
 
 var _ Buffer = (*blockingBuffer)(nil)
 
+type BlockingBufferOption func(b *blockingBuffer)
+
+func WithBlockingBufferTestingKnobs(knobs BlockingBufferTestingKnobs) BlockingBufferOption {
+	return func(b *blockingBuffer) {
+		b.knobs = knobs
+	}
+}
+
 func (b *blockingBuffer) pop() (e Event, ok bool, err error) {
+	if b.knobs.BeforePop != nil {
+		b.knobs.BeforePop()
+	}
+
 	b.mu.Lock()
 	defer b.mu.Unlock()
+
 	if b.mu.closed {
 		return Event{}, false, ErrBufferClosed{reason: b.mu.reason}
 	}
@@ -143,6 +166,7 @@ func (b *blockingBuffer) pop() (e Event, ok bool, err error) {
 		close(b.mu.drainCh)
 		b.mu.drainCh = nil
 	}
+
 	return e, ok, nil
 }
 
@@ -254,6 +278,10 @@ func (b *blockingBuffer) AcquireMemory(ctx context.Context, n int64) (alloc Allo
 
 // Add implements Writer interface.
 func (b *blockingBuffer) Add(ctx context.Context, e Event) error {
+	if b.knobs.BeforeAdd != nil {
+		ctx, e = b.knobs.BeforeAdd(ctx, e)
+	}
+
 	if log.V(2) {
 		log.Infof(ctx, "Add event: %s", e.String())
 	}
@@ -290,7 +318,16 @@ func (b *blockingBuffer) tryDrain() chan struct{} {
 }
 
 // Drain implements Writer interface.
-func (b *blockingBuffer) Drain(ctx context.Context) error {
+func (b *blockingBuffer) Drain(ctx context.Context) (err error) {
+	if b.knobs.BeforeDrain != nil {
+		ctx = b.knobs.BeforeDrain(ctx)
+	}
+	if b.knobs.AfterDrain != nil {
+		defer func() {
+			b.knobs.AfterDrain(err)
+		}()
+	}
+
 	if drained := b.tryDrain(); drained != nil {
 		select {
 		case <-ctx.Done():
@@ -304,7 +341,13 @@ func (b *blockingBuffer) Drain(ctx context.Context) error {
 }
 
 // CloseWithReason implements Writer interface.
-func (b *blockingBuffer) CloseWithReason(ctx context.Context, reason error) error {
+func (b *blockingBuffer) CloseWithReason(ctx context.Context, reason error) (err error) {
+	if b.knobs.AfterCloseWithReason != nil {
+		defer func() {
+			b.knobs.AfterCloseWithReason(err)
+		}()
+	}
+
 	// Close quota pool -- any requests waiting to acquire will receive an error.
 	b.qp.Close("blocking buffer closing")
 

--- a/pkg/ccl/changefeedccl/kvevent/testing_knobs.go
+++ b/pkg/ccl/changefeedccl/kvevent/testing_knobs.go
@@ -1,0 +1,17 @@
+// Copyright 2025 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package kvevent
+
+import "context"
+
+// BlockingBufferTestingKnobs are testing knobs for blocking buffers.
+type BlockingBufferTestingKnobs struct {
+	BeforeAdd            func(ctx context.Context, e Event) (context.Context, Event)
+	BeforePop            func()
+	BeforeDrain          func(ctx context.Context) context.Context
+	AfterDrain           func(err error)
+	AfterCloseWithReason func(err error)
+}

--- a/pkg/ccl/changefeedccl/kvfeed/kv_feed.go
+++ b/pkg/ccl/changefeedccl/kvfeed/kv_feed.go
@@ -172,18 +172,18 @@ func Run(ctx context.Context, cfg Config) error {
 	// Regardless of whether drain succeeds, we must also close the buffer to release
 	// any resources, and to let the consumer (changeAggregator) know that no more writes
 	// are expected so that it can transition to a draining state.
-	err = errors.CombineErrors(
-		f.writer.Drain(ctx),
-		f.writer.CloseWithReason(ctx, kvevent.ErrNormalRestartReason),
-	)
-
-	if err == nil {
-		// This context is canceled by the change aggregator when it receives
-		// an error reading from the Writer that was closed above.
-		<-ctx.Done()
+	if err := f.writer.Drain(ctx); err != nil {
+		err := errors.Wrap(err, "failed to drain kv feed writer")
+		return errors.CombineErrors(err, f.writer.CloseWithReason(ctx, err))
+	}
+	if err := f.writer.CloseWithReason(ctx, kvevent.ErrNormalRestartReason); err != nil {
+		return err
 	}
 
-	return err
+	// This context is canceled by the change aggregator when it receives
+	// an error reading from the Writer that was closed above.
+	<-ctx.Done()
+	return ctx.Err()
 }
 
 func startLaggingRangesObserver(

--- a/pkg/ccl/changefeedccl/testing_knobs.go
+++ b/pkg/ccl/changefeedccl/testing_knobs.go
@@ -8,6 +8,7 @@ package changefeedccl
 import (
 	"context"
 
+	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvevent"
 	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/kvfeed"
 	"github.com/cockroachdb/cockroach/pkg/ccl/changefeedccl/resolvedspan"
 	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
@@ -111,6 +112,10 @@ type TestingKnobs struct {
 
 	// OverrideCursorAge is used to change how old a cursor is. Returns time in nanoseconds.
 	OverrideCursorAge func() int64
+
+	// MakeKVFeedToAggregatorBufferKnobs is used to make a fresh set of testing knobs
+	// to pass to the constructor of the kv feed to change aggregator buffer.
+	MakeKVFeedToAggregatorBufferKnobs func() kvevent.BlockingBufferTestingKnobs
 }
 
 // ModuleTestingKnobs is part of the base.ModuleTestingKnobs interface.

